IGNITE-16969 Add developer documentation to REST modules (#1028)
apache/ignite-3: modules/rest/README.md

# ignite-rest

This module defines [RestComponent](src/main/java/org/apache/ignite/internal/rest/RestComponent.java) that is responsible for:

> - at compile time it aggregates all REST API definitions by `@OpenAPIInclude` annotation
> - at runtime, it creates a micronaut context, injects all needed beans into the context
> - starts the micronaut server at the configured port range

Reviewer: ```suggestion
- aggregating all REST API definitions by `@OpenAPIInclude` annotation at compile time
- creating a micronaut context at runtime and injecting all needed beans into the context
- starting the micronaut server at the configured port range
```

> During the build time, the [Open API spec]() is generated from all API definitions that are included in

Reviewer: Empty link

`@OpenAPIInclude` annotation. The spec is located in `ignite-rest/openapi`.

## How to create a new REST Endpoint

- define API in `ignite-rest-api` module
- link the API interface in `RestComponent`'s `@OpenAPIInclude` section
- implement an API interface in the module that is responsible for the API scope
- define micronaut factory that is responsible for the creation of beans that are needed for your Controller
- for extending the exception handling mechanism, implement micronaut's `ExceptionHandler` and include it in the micronaut factory

> - create the instance of the RestFactory in IgniteImpl constructor and put the factory to the RestComponent's constructor

Reviewer: ```suggestion
- create the instance of the RestFactory in the IgniteImpl constructor and put the factory to the RestComponent's constructor
```

> The class that implements the API should be a valid micronaut controller.
> Module with the API implementation should configure `micronaut-inject-java` annotation processor.

> Micronaut factories have to implement `RestFactory` from `ignite-rest-api` 

## Example of REST Endpoint

> - API definition [ClusterManagementApi](../rest-api/src/main/java/org/apache/ignite/internal/rest/api/cluster)

Reviewer: ```suggestion
- API definition [ClusterManagementApi](../rest-api/src/main/java/org/apache/ignite/internal/rest/api/cluster/ClusterManagementApi.java)
```
